README.md
# Side Menu Animation

[![pub package](https://img.shields.io/pub/v/side_menu_animation.svg)](https://pub.dartlang.org/packages/side_menu_animation)

Animated Side Menu with customizable UI. Inspired in Yalantis library (Android/iOS).
Original Design: https://dribbble.com/shots/1689922-Side-Menu-Animation

## Sample

<TABLE BORDER style="margin-left: auto;
  margin-right: auto;">
    <TR>
        <TH style="text-align:center">SIDE MENU ANIMATION</TH>
        <TH style="text-align:center">SIDE MENU ANIMATION - DRAG GESTURE</TH>
    </TR>
	<TR>
		<TD><img width="300" height="600" src="https://media.giphy.com/media/UScpVNqe11bqSfpQ6G/giphy.gif"></TD> 
        <TD><img width="300" height="600" src="https://media.giphy.com/media/YdiRzuOlK8bw7ckaRc/giphy-downsized-large.gif"></TD>
	</TR>
</TABLE>

## Learn how to build this package

- Video 1: https://www.youtube.com/watch?v=vcdETKdI15E
- Video 2: https://www.youtube.com/watch?v=W7mxTcwX5Wg

(Don't forget to subscribe and like)

## Features
- Optional parameter to tap outside to dismiss
- Optional parameter to change the scrimColor
- Custom color item for the side menu selected
- Custom color item for the side menu unselected
- Custom width for the side menu
- Custom duration of the animation side menu
- Display the menu from left or right

## Getting Started

You should ensure that you add the router as a dependency in your flutter project.

```yaml
dependencies:
 side_menu_animation: "^0.0.1"
```

You should then run `flutter packages upgrade` or update your packages in IntelliJ.

## Example Project

There is a example project in the `example` folder. Check it out. Otherwise, keep reading to get up and running.

## Usage

Need to include the import the package to the dart file where it will be used, use the below command,

```dart
import 'package:side_menu_animation/side_menu_animation.dart';
```

## SideMenu

```dart
class SideMenuScreen extends StatelessWidget {
  final _index = ValueNotifier<int>(1);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SideMenuAnimation(
        appBarBuilder: (showMenu) => AppBar(
          leading: IconButton(icon: Icon(Icons.menu, color: Colors.black), onPressed: showMenu),
          backgroundColor: Theme.of(context).scaffoldBackgroundColor,
          elevation: 0,
          brightness: Brightness.light,
          centerTitle: true,
          title: ValueListenableBuilder<int>(
            valueListenable: _index,
            builder: (_, value, __) => Text(value.toString(), style: TextStyle(color: Colors.black)),
          ),
        ),
        views: [YourCustomViews1Here(), YourCustomViews2Here()],
        items: [MyCustomItem1Here(), MyCustomItem2Here()],
        selectedColor: Color(0xFFFF595E),
        unselectedColor: Color(0xFF1F2041),
        tapOutsideToDismiss: true,
        scrimColor: Colors.black45,
        onItemSelected: (value) {
          if (value > 0 && value != _index.value) _index.value = value;
        },
      ),
    );
  }
}
```

## SideMenuBuilder

```dart
class SideMenuBuilderScreen extends StatelessWidget {
  final _index = ValueNotifier<int>(1);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SideMenuAnimation.builder(
        builder: (showMenu) {
          return Scaffold(
            appBar: AppBar(
              leading: IconButton(icon: Icon(Icons.menu, color: Colors.black), onPressed: showMenu),
              backgroundColor: Theme.of(context).scaffoldBackgroundColor,
              elevation: 0,
              brightness: Brightness.light,
              centerTitle: true,
              title: ValueListenableBuilder<int>(
                valueListenable: _index,
                builder: (_, value, __) => Text(_value.toString(), style: TextStyle(color: Colors.black)),
              ),
            ),
            body: ValueListenableBuilder<int>(
              valueListenable: _index,
              builder: (_, value, __) => IndexedStack(
                index: value - 1,
                children: [YourCustomViews1Here(), YourCustomViews2Here()],
              ),
            ),
          );
        },
       items: [MyCustomItem1Here(), MyCustomItem2Here()],
        selectedColor: Color(0xFFFF595E),
        unselectedColor: Color(0xFF1F2041),
        onItemSelected: (value) {
          if (value > 0 && value != _index.value) _index.value = value;
        },
      ),
    );
  }
}
```

For more info about the ussage, check the `example` project.

## Contact

You can follow me on twitter [@diegoveloper](https://www.twitter.com/diegoveloper) , [Youtube channel](https://www.youtube.com/diegoveloper)

If you want to contribute with the project, just open a Pull request :), all contributions are welcome.


## Contribution

If you want to contribute with this package, follow this steps:

- Fork this repository.
- Do your changes! You can add your name with link into the `CONTRIBUTORS.md` file.
- Before push your changes run `dartfmt . -w`.
- Create a Pull request on Github from your fork/branch to my repo(main branch).

CHANGELOG.md
## [1.1.0]

* Official release.

## [1.0.0-beta]

* Null safety migration. Thanks Brayan Cantos

## [0.0.2]

* `enableEdgeDragGesture` and `edgeDragWidth` was added. Thanks Brayan Cantos.
* `position` property and `SideMenuPosition` were added, now we can display the menu from the left or right. Thanks Jorge Alejandro.

## [0.0.1 - 0.0.1+2]

* Initial version.


test/side_menu_animation_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:side_menu_animation/side_menu_animation.dart';

void main() {
  group('SideMenuAnimation::Tests', () {
    testWidgets('SideMenuAnimation:: Press menu - Show Scrim Widget',
        (tester) async {
      await tester.pumpWidget(MaterialApp(
        home: Scaffold(
          body: SideMenuAnimation(
            appBarBuilder: (showMenu) => AppBar(
              leading: IconButton(
                  icon: const Icon(Icons.menu, color: Colors.black),
                  onPressed: showMenu),
            ),
            views: List.generate(4, (index) => Container()),
            items: List.generate(4, (index) => const SizedBox())
                .map((value) => value)
                .toList(),
            tapOutsideToDismiss: true,
            onItemSelected: (value) {},
          ),
        ),
      ));

      await tester.tap(find.byIcon(Icons.menu));
      await tester.pumpAndSettle();

      expect(find.byType(AnimatedContainer), findsOneWidget);
    });

    testWidgets(
        'SideMenuAnimation:: Press menu, press scrim widget and hide menu',
        (tester) async {
      await tester.pumpWidget(MaterialApp(
        home: Scaffold(
          body: SideMenuAnimation(
            appBarBuilder: (showMenu) => AppBar(
              leading: IconButton(
                  icon: const Icon(Icons.menu, color: Colors.black),
                  onPressed: showMenu),
            ),
            views: List.generate(4, (index) => Container()),
            items: List.generate(4, (index) => const SizedBox())
                .map((value) => value)
                .toList(),
            tapOutsideToDismiss: true,
            onItemSelected: (value) {},
          ),
        ),
      ));

      await tester.tap(find.byIcon(Icons.menu));
      await tester.pumpAndSettle();
      expect(find.byType(AnimatedContainer), findsOneWidget);
      await tester.tap(find.byType(AnimatedContainer));
      await tester.pumpAndSettle();
      expect(find.byType(AnimatedContainer), findsNothing);
    });

    testWidgets(
        'SideMenuAnimation:: Press menu, press item 2 and display page 2',
        (tester) async {
      int? _indexSelected;
      await tester.pumpWidget(MaterialApp(
        home: Scaffold(
          body: SideMenuAnimation(
            appBarBuilder: (showMenu) => AppBar(
              leading: IconButton(
                  icon: const Icon(Icons.menu, color: Colors.black),
                  onPressed: showMenu),
            ),
            views: List.generate(
                4,
                (index) => Container(
                    color: Colors.red,
                    child: Center(child: Text('Page ${index + 1}')))),
            items: List.generate(
              4,
              (index) => Center(
                child: Text('Item $index'),
              ),
            ).map((value) => value).toList(),
            tapOutsideToDismiss: true,
            scrimColor: Colors.black45,
            onItemSelected: (value) {
              _indexSelected = value;
            },
          ),
        ),
      ));
      await tester.tap(find.byIcon(Icons.menu));
      await tester.pumpAndSettle();
      await tester.tap(find.text('Item 3'));
      await tester.pumpAndSettle();
      expect(find.text('Page 3'), findsOneWidget);
      expect(_indexSelected, 3);
    });
  });
}


example/lib/main.dart
import 'package:example/sidemenu_builder_screen.dart';
import 'package:example/sidemenu_screen.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    SystemChrome.setSystemUIOverlayStyle(
      SystemUiOverlayStyle.dark.copyWith(statusBarColor: Colors.transparent),
    );
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'SideMenuAnimation-Example',
      home: SampleScreen(),
    );
  }
}

class SampleScreen extends StatelessWidget {
  void _goToExample(BuildContext context, Widget newPage) =>
      Navigator.push(context, MaterialPageRoute(builder: (_) => newPage));

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('SideMenuAnimation-Example'),
      ),
      body: Padding(
        padding: const EdgeInsets.all(15),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            CupertinoButton(
              color: Colors.blueAccent,
              pressedOpacity: .75,
              child: Text('SideMenuAnimation'),
              onPressed: () => _goToExample(context, SideMenuScreen()),
            ),
            const SizedBox(height: 20),
            CupertinoButton(
              color: Colors.blueAccent,
              pressedOpacity: .75,
              child: Text('SideMenuAnimation.builder'),
              onPressed: () => _goToExample(context, SideMenuBuilderScreen()),
            ),
          ],
        ),
      ),
      drawer: Drawer(),
    );
  }
}


example/lib/screen.dart
import 'package:example/menu_value.dart';
import 'package:flutter/material.dart';

class Screen extends StatelessWidget {
  const Screen({Key key, @required this.itemsScreen}) : super(key: key);

  final List<MenuValues> itemsScreen;

  @override
  Widget build(BuildContext context) {
    return Column(
      children: List.generate(
        itemsScreen.length,
        (index) => Expanded(
          child: Container(
            color: itemsScreen[index].color,
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.stretch,
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                Icon(itemsScreen[index].icon, size: 75, color: Colors.white),
                const SizedBox(height: 10),
                Text(
                  itemsScreen[index].title,
                  textAlign: TextAlign.center,
                  style: TextStyle(
                      color: Colors.white,
                      fontSize: 25,
                      fontWeight: FontWeight.w700),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}


example/lib/menu_value.dart
import 'package:flutter/material.dart';

class MenuValues {
  const MenuValues({this.title, this.icon, this.items, this.color});
  final String title;
  final IconData icon;
  final Color color;
  final List<MenuValues> items;
}

const myMenuValue = const [
  MenuValues(icon: Icons.close),
  MenuValues(
    icon: Icons.music_note_rounded,
    title: 'Music',
    items: const [
      MenuValues(
          icon: Icons.music_note, title: 'Songs', color: Color(0xFF5863F8)),
      MenuValues(
          icon: Icons.play_arrow,
          title: 'Now Playing',
          color: Color(0xFFFF3366)),
      MenuValues(icon: Icons.album, title: 'Albums', color: Color(0xFFFFE433)),
    ],
  ),
  MenuValues(
    icon: Icons.phone_bluetooth_speaker_rounded,
    title: 'Calls',
    items: const [
      MenuValues(
          icon: Icons.phone_callback_rounded,
          title: 'Incoming',
          color: Color(0xFF2CDA9D)),
      MenuValues(
          icon: Icons.phone_missed_rounded,
          title: 'Missing',
          color: Color(0xFF7678ED)),
      MenuValues(
          icon: Icons.phone_disabled_rounded,
          title: 'Outgoing ',
          color: Color(0xFF446DF6)),
    ],
  ),
  MenuValues(
    icon: Icons.cloud,
    title: 'Cloud',
    items: const [
      MenuValues(
          icon: Icons.download_rounded,
          title: 'Downloading',
          color: Color(0xFFFF4669)),
      MenuValues(
          icon: Icons.upload_file, title: 'Done', color: Color(0xFFFF69EB)),
      MenuValues(
          icon: Icons.cloud_upload, title: 'Upload', color: Color(0xFF2CDA9D)),
    ],
  ),
  MenuValues(
    icon: Icons.wifi,
    title: 'Wifi',
    items: const [
      MenuValues(
          icon: Icons.wifi_off_rounded, title: 'Off', color: Color(0xFF5AD2F4)),
      MenuValues(
          icon: Icons.signal_wifi_4_bar_lock_sharp,
          title: 'Lock',
          color: Color(0xFFFF3366)),
      MenuValues(
          icon: Icons.perm_scan_wifi_rounded,
          title: 'Limit',
          color: Color(0xFFFFC07F)),
    ],
  ),
  MenuValues(
    icon: Icons.favorite,
    title: 'Favorites',
    items: const [
      MenuValues(
          icon: Icons.favorite, title: 'Favorite', color: Color(0xFF5863F8)),
      MenuValues(
          icon: Icons.favorite_border,
          title: 'Not Favorite',
          color: Color(0xFFF7C548)),
      MenuValues(
          icon: Icons.volunteer_activism,
          title: 'Activism',
          color: Color(0xFF00A878)),
    ],
  ),
  MenuValues(
    icon: Icons.network_cell,
    title: 'Networks',
    items: const [
      MenuValues(icon: Icons.wifi, title: 'Wifi', color: Color(0xFF96858F)),
      MenuValues(
          icon: Icons.network_cell, title: 'Network', color: Color(0xFF6D7993)),
      MenuValues(
          icon: Icons.bluetooth, title: 'Bluetooth', color: Color(0xFF9099A2)),
    ],
  ),
];


lib/side_menu_animation.dart
import 'package:flutter/material.dart';

import 'src/circular_apearing.dart';

/// Signature for creating a widget with a `showMenu` callback
/// for opening the Side Menu.
///
/// See also:
/// * [SideMenuAnimation.builder]
/// * [SideMenuAnimationAppBarBuilder]
typedef SideMenuAnimationBuilder = Widget Function(VoidCallback showMenu);

/// Signature for creating an [AppBar] widget with a
/// `showMenu` callback for opening the Side Menu.
///
/// See also:
/// * [SideMenuAnimation].
typedef SideMenuAnimationAppBarBuilder = AppBar Function(VoidCallback showMenu);

const _sideMenuWidth = 100.0;
const _sideMenuDuration = Duration(milliseconds: 800);
const _kEdgeDragWidth = 20.0;

/// # SideMenuPosition
/// This enum is the position selector of the menu.
///
/// {@template SideMenuPosition.right}
/// ## right
/// Set the position of the menu in the right side on the screen
/// {@endtemplate}
///
/// {@template SideMenuPosition.left}
/// ## left
/// Set the position of the menu in the left side on the screen
/// {@endtemplate}
enum SideMenuPosition {
  /// {@macro SideMenuPosition.right}
  right,

  /// {@macro SideMenuPosition.left}
  left,
}

extension _SideMenuPositionX on SideMenuPosition {
  bool get isLeft => this == SideMenuPosition.left;
  bool get isRight => this == SideMenuPosition.right;
}

/// The [SideMenuAnimation] controls the items from the lateral menu
/// and also can control the circular reveal transition.
class SideMenuAnimation extends StatefulWidget {
  /// Creates a [SideMenuAnimation] without Circular Reveal animation.
  /// Also it is responsible for updating/changing the [AppBar]
  /// based on the index we receive.
  const SideMenuAnimation.builder({
    Key? key,
    required this.builder,
    required this.items,
    required this.onItemSelected,
    this.position = SideMenuPosition.left,
    this.selectedColor = Colors.black,
    this.unselectedColor = Colors.green,
    double? menuWidth,
    Duration? duration,
    this.tapOutsideToDismiss = false,
    this.scrimColor = Colors.transparent,
    double? edgeDragWidth,
    this.enableEdgeDragGesture = false,
    this.curveAnimation = Curves.linear,
  })  : views = null,
        appBarBuilder = null,
        indexSelected = null,
        menuWidth = menuWidth ?? _sideMenuWidth,
        duration = duration ?? _sideMenuDuration,
        edgeDragWidth = edgeDragWidth ?? _kEdgeDragWidth,
        super(key: key);

  /// Creates a [SideMenuAnimation] with Circular Reveal animation.
  /// Also it is responsible for updating/changing the [AppBar]
  /// based on the index we receive.
  const SideMenuAnimation({
    Key? key,
    required this.views,
    required this.items,
    required this.onItemSelected,
    this.position = SideMenuPosition.left,
    this.selectedColor = Colors.black,
    this.unselectedColor = Colors.green,
    double? menuWidth,
    Duration? duration,
    this.appBarBuilder,
    this.indexSelected = 0,
    this.tapOutsideToDismiss = false,
    this.scrimColor = Colors.transparent,
    double? edgeDragWidth,
    this.enableEdgeDragGesture = false,
    this.curveAnimation = Curves.linear,
  })  : builder = null,
        menuWidth = menuWidth ?? _sideMenuWidth,
        duration = duration ?? _sideMenuDuration,
        edgeDragWidth = edgeDragWidth ?? _kEdgeDragWidth,
        super(key: key);

  /// `builder` builds a view/page based on the `selectedIndex.
  /// It also comes with a `showMenu` callback for opening the Side Menu.
  final SideMenuAnimationBuilder? builder;

  /// `appBarBuilder` returns an [AppBar] based on the `selectedIndex`
  /// It also comes with the `showMenu` callback
  /// where we can use to open the Side Menu.
  final SideMenuAnimationAppBarBuilder? appBarBuilder;

  /// List of items that we want to display on the Side Menu.
  final List<Widget> items;

  /// Function where we receive the current index selected.
  final ValueChanged<int> onItemSelected;

  /// [Color] used for the background of the selected item.
  final Color selectedColor;

  /// [Color] used for the background of the unselected item.
  final Color unselectedColor;

  /// Menu width for the Side Menu.
  final double menuWidth;

  /// Duration for the animation when the menu appears, this is
  /// the total duration, each item has total_duration/items.lenght
  final Duration duration;

  /// Pages/Views we pass to the widge to display with a circular
  /// reveal animation
  final List<Widget>? views;

  /// Initial index selected
  final int? indexSelected;

  /// Enables to dismiss the [SideMenuAnimation] when user taps outside
  /// the widget.
  /// It's `false` by default.
  final bool tapOutsideToDismiss;

  /// Defines the `position` of the sider menu: `left` of `right`.
  /// By default it is [SideMenuPosition.left].
  final SideMenuPosition position;

  /// If `tapOutsideToDismiss` is true, then the `scrimColor` is enabled
  /// to change, this is the panel where we tap to dismiss the Side Menu.
  final Color scrimColor;

  /// Enables swipe from left to right to display the menu,
  /// it's `false` by default.
  final bool enableEdgeDragGesture;

  /// If `enableEdgeDragGesture` is true, then we can change
  /// the `edgeDragWidth`, this is the width of the area where we do swipe.
  final double edgeDragWidth;

  /// [Curve] used for the animation
  final Curve curveAnimation;

  @override
  _SideMenuAnimationState createState() => _SideMenuAnimationState();
}

class _SideMenuAnimationState extends State<SideMenuAnimation>
    with SingleTickerProviderStateMixin {
  late AnimationController _animationController;

  late int _selectedIndex;
  late int _oldSelectedIndex;
  int _selectedColor = 1;

  late ColorTween _scrimColorTween;

  @override
  void initState() {
    _selectedIndex = widget.indexSelected ?? 0;
    _oldSelectedIndex = _selectedIndex;
    _animationController =
        AnimationController(vsync: this, duration: widget.duration);
    _animationController.forward(from: 1.0);
    _createColorTween();
    super.initState();
  }

  void _createColorTween() {
    _scrimColorTween = ColorTween(
      end: Colors.transparent,
      begin: widget.scrimColor,
    );
  }

  @override
  void didUpdateWidget(SideMenuAnimation oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (oldWidget.scrimColor != widget.scrimColor) _createColorTween();
    if (oldWidget.duration != widget.duration) {
      _animationController.duration = widget.duration;
    }
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  void _displayMenuDragGesture(DragEndDetails endDetails) {
    final velocity = endDetails.primaryVelocity!;
    if (widget.position.isLeft) {
      if (velocity > 0) _animationReverse();
    } else {
      if (velocity < 0) _animationReverse();
    }
  }

  void _animationReverse() {
    _animationController.reverse();
  }

  @override
  Widget build(BuildContext context) {
    return Material(
      child: LayoutBuilder(
        builder: (context, constraints) {
          final itemSize = constraints.maxHeight / widget.items.length;
          return Stack(
            children: [
              if (widget.builder != null) widget.builder!(_animationReverse),
              if (widget.appBarBuilder != null)
                Scaffold(
                  appBar: widget.appBarBuilder!(_animationReverse),
                  body: CircularApearing(
                    animationController: _animationController,
                    dx: widget.position.isLeft ? 0.0 : constraints.maxWidth,
                    dy: (itemSize * _selectedIndex) + (itemSize / 2),
                    child: widget.views![_oldSelectedIndex],
                    enteringChild: widget.views![_selectedIndex],
                  ),
                ),
              AnimatedBuilder(
                animation: _animationController,
                builder: (context, child) => Stack(
                  children: [
                    if (widget.tapOutsideToDismiss &&
                        _animationController.value < 1)
                      Align(
                        child: GestureDetector(
                          onTap: () => _animationController.forward(from: 0.0),
                          child: AnimatedContainer(
                            duration: widget.duration,
                            color: _scrimColorTween.evaluate(
                              Tween(begin: 0.0, end: 1.0)
                                  .animate(_animationController),
                            ),
                          ),
                        ),
                      ),
                    if (widget.enableEdgeDragGesture &&
                        _animationController.isCompleted)
                      Align(
                        alignment: widget.position.isLeft
                            ? Alignment.centerLeft
                            : Alignment.centerRight,
                        child: GestureDetector(
                          onHorizontalDragEnd: _displayMenuDragGesture,
                          behavior: HitTestBehavior.translucent,
                          excludeFromSemantics: true,
                          child: Container(width: widget.edgeDragWidth),
                        ),
                      ),
                    for (int i = 0; i < widget.items.length; i++)
                      MenuItem(
                        position: widget.position,
                        index: i,
                        length: widget.items.length,
                        width: widget.menuWidth,
                        height: itemSize,
                        controller: _animationController,
                        curve: widget.curveAnimation,
                        color: (i == _selectedColor)
                            ? widget.selectedColor
                            : widget.unselectedColor,
                        onTap: () {
                          if (i != _selectedColor) {
                            if (i != 0) {
                              setState(() {
                                _oldSelectedIndex = _selectedIndex;
                                _selectedIndex = i - 1;
                                _selectedColor = i;
                              });
                            }
                            widget.onItemSelected(i);
                          }
                        },
                        child: widget.items[i],
                      ),
                  ],
                ),
              ),
            ],
          );
        },
      ),
    );
  }
}

/// {@template MenuItem}
/// A [MenuItem] for the [SideMenuAnimation]
/// {@endtemplate}
class MenuItem extends StatelessWidget {
  /// {@macro MenuItem}
  const MenuItem({
    Key? key,
    required this.position,
    required this.index,
    required this.length,
    required this.width,
    required this.height,
    required this.child,
    required this.curve,
    required this.controller,
    required this.color,
    required this.onTap,
  }) : super(key: key);

  /// `position` for the menu
  final SideMenuPosition position;

  /// `index` for the [MenuItem]
  final int index;

  /// Number of items
  final int length;

  /// `width` for the [MenuItem]
  final double width;

  /// `height` for the [MenuItem]
  final double height;

  /// widget `child`
  final Widget child;

  /// [AnimationController] used in the [SideMenuAnimation]
  final AnimationController controller;

  /// Animation [Curve]
  final Curve curve;

  /// Background `color`
  final Color color;

  /// Callback invoked `onTap`
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    final _intervalGap = 1 / length;
    final _index = controller.status == AnimationStatus.reverse
        ? length - 1 - index
        : index;
    final _animation = Tween(begin: 0.0, end: 1.6).animate(
      CurvedAnimation(
        parent: controller,
        curve: Interval(
          _intervalGap * _index,
          _intervalGap * (_index + 1),
          curve: curve,
        ),
      ),
    );
    return Positioned(
      left: position.isLeft ? 0 : null,
      right: position.isRight ? 0 : null,
      top: height * index,
      width: width,
      height: height,
      child: Transform(
        transform: Matrix4.identity()
          ..setEntry(3, 2, 0.001)
          ..rotateY(-_animation.value),
        alignment: position.isRight ? Alignment.topRight : Alignment.topLeft,
        child: Material(
          color: color,
          child: InkWell(
            onTap: () {
              controller.forward(from: 0.0);
              onTap();
            },
            child: child,
          ),
        ),
      ),
    );
  }
}


lib/src/circular_apearing.dart
import 'package:flutter/widgets.dart';

/// {@template CircularApearing}
/// Creates an animation where a new widget apears in from of another one
/// as circle with increasing radius with a center positioned in (`dx`, `dy`).
/// {@endtemplate}
class CircularApearing extends StatefulWidget {
  /// {@macro CircularApearing}
  const CircularApearing({
    Key? key,
    required this.child,
    this.enteringChild,
    required this.dx,
    required this.dy,
    required this.animationController,
  }) : super(key: key);

  /// [Widget] `child` that is currently shown
  final Widget child;

  /// `enteringChild` with the [CircularApearing] animation
  final Widget? enteringChild;

  /// Initial position in the X-axis for the [CircularApearing] animation
  final double dx;

  /// Initial position in the Y-axis for the [CircularApearing] animation
  final double dy;

  /// `animationcontroller` for [CircularApearing] animation
  final AnimationController animationController;

  @override
  _CircularApearingState createState() => _CircularApearingState();
}

class _CircularApearingState extends State<CircularApearing> {
  late Widget child;
  Widget? enteringChild;

  @override
  void initState() {
    super.initState();
    child = widget.child;
    enteringChild = widget.enteringChild;
    widget.animationController.addStatusListener((status) {
      if (status == AnimationStatus.completed && enteringChild != null) {
        setState(() {
          child = enteringChild!;
          enteringChild = null;
        });
      }
    });
  }

  @override
  void didUpdateWidget(covariant CircularApearing oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (oldWidget.child != widget.child ||
        oldWidget.enteringChild != widget.enteringChild) {
      child = widget.child;
      enteringChild = widget.enteringChild;
    }
  }

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [
        child,
        if (enteringChild != null)
          ClipPath(
            clipper: _CircularApearingClipper(
              end: 3.0,
              dx: widget.dx,
              dy: widget.dy,
              controller: widget.animationController,
            ),
            child: enteringChild,
          ),
      ],
    );
  }
}

class _CircularApearingClipper extends CustomClipper<Path> {
  _CircularApearingClipper({
    required this.end,
    required this.dx,
    required this.dy,
    required this.controller,
  })   : assert(end > 0.0),
        super(reclip: controller);

  final double end;
  final double dx, dy;
  final AnimationController controller;

  @override
  Path getClip(Size size) {
    final percent = controller.status == AnimationStatus.forward
        ? Tween(begin: 0.0, end: 3.0).animate(controller).value
        : 3.0;
    return Path()
      ..addOval(
        Rect.fromCenter(
          center: Offset(dx, dy),
          width: size.width * percent,
          height: size.height * percent,
        ),
      );
  }

  @override
  bool shouldReclip(covariant _CircularApearingClipper oldClipper) =>
      oldClipper.end != end;
}


example/lib/sidemenu_screen.dart
import 'package:example/menu_value.dart';
import 'package:example/screen.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:side_menu_animation/side_menu_animation.dart';

class SideMenuScreen extends StatelessWidget {
  final _index = ValueNotifier<int>(1);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SideMenuAnimation(
        appBarBuilder: (showMenu) => AppBar(
          leading: IconButton(
              icon: Icon(Icons.menu, color: Colors.black), onPressed: showMenu),
          backgroundColor: Theme.of(context).scaffoldBackgroundColor,
          elevation: 0,
          systemOverlayStyle: SystemUiOverlayStyle.light,
          centerTitle: true,
          title: ValueListenableBuilder<int>(
            valueListenable: _index,
            builder: (_, value, __) => Text(myMenuValue[value].title,
                style: TextStyle(color: Colors.black)),
          ),
        ),
        views: List.generate(
          myMenuValue.length - 1,
          (index) => Screen(itemsScreen: myMenuValue[index + 1].items),
        ),
        position: SideMenuPosition.right,
        items: myMenuValue
            .map((value) => Icon(value.icon, color: Colors.white, size: 50))
            .toList(),
        selectedColor: Color(0xFFFF595E),
        unselectedColor: Color(0xFF1F2041),
        tapOutsideToDismiss: true,
        scrimColor: Colors.black45,
        onItemSelected: (value) {
          if (value > 0 && value != _index.value) _index.value = value;
        },
      ),
    );
  }
}


example/lib/sidemenu_builder_screen.dart
import 'package:example/menu_value.dart';
import 'package:example/screen.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:side_menu_animation/side_menu_animation.dart';

class SideMenuBuilderScreen extends StatelessWidget {
  final _index = ValueNotifier<int>(1);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SideMenuAnimation.builder(
        builder: (showMenu) {
          return Scaffold(
            appBar: AppBar(
              leading: IconButton(
                  icon: Icon(Icons.menu, color: Colors.black),
                  onPressed: showMenu),
              backgroundColor: Theme.of(context).scaffoldBackgroundColor,
              elevation: 0,
              systemOverlayStyle: SystemUiOverlayStyle.light,
              centerTitle: true,
              title: ValueListenableBuilder<int>(
                valueListenable: _index,
                builder: (_, value, __) => Text(myMenuValue[value].title,
                    style: TextStyle(color: Colors.black)),
              ),
            ),
            body: ValueListenableBuilder<int>(
              valueListenable: _index,
              builder: (_, value, __) => IndexedStack(
                index: value - 1,
                children: List.generate(
                  myMenuValue.length - 1,
                  (index) => Screen(itemsScreen: myMenuValue[index + 1].items),
                ),
              ),
            ),
          );
        },
        enableEdgeDragGesture: true,
        items: myMenuValue
            .map((value) => Icon(value.icon, color: Colors.white, size: 50))
            .toList(),
        selectedColor: Color(0xFFFF595E),
        unselectedColor: Color(0xFF1F2041),
        onItemSelected: (value) {
          if (value > 0 && value != _index.value) _index.value = value;
        },
      ),
    );
  }
}
